### ISSUE TITLE
Current State of JIT

### ISSUE THREAD

**Issue by WilliamLCobb** [WilliamLCobb/iNDS#44]
In order to keep this organized, I'll update this comment with any progress on JIT.

So far JIT code is successfully compiled and running on iOS 8.3 on a 32bit processor. I'm sure iOS 9 on a 32 bit processor would also work fine I just don't want to upgrade my iPad to test it. Inside armInnerLoop(), JIT code will run anywhere from 1 to 100 time before it crashes. The function where JIT execution is compiled and then run is cpuExecuteLJIT() inside ArmLJit.cpp.

The main problem we're facing is the code in the JIT block branches to many other functions inside ArmLJit.cpp. These functions then push the callee registers r4-r8. Oddly, these callee saved registers are not pushed back onto the stack before returning to the JIT block as they should be. This causes bad access errors and the crashing we currently see.

**Comment by pmp174** [comment]
I wish I had your skills but that's awesome.

**Comment by WilliamLCobb** [comment]
Haha I've learned almost all of this in the past year. Working on projects like this just teaches you a ton really quickly

**Comment by milch** [comment]
Looking at the source, the relevant sections are in `RegisterMapImp::StartBlock` and `RegisterMapImp::EndBlock`. The first one calls out to `jit_prolog` (which is actually implemented in `arm_prolog`) and the second one calls out to `jit_ret` (implemented in `arm_ret`). Both pushes and pops are handled in those functions.

**Comment by WilliamLCobb** [comment]
Thanks @milch I think those write the push and pops at the beginning and end of the JIT block though and seem working okay. 

For anyone trying to debug the JIT generated code you can change EMU_setCpuMode in emulatorViewController.mm from 1 to 2 to turn on JIT  and set a breakpoint at the end of armExecuteLJIT(). Then view the memory at opfun by right clicking opfun and click view memory and that's the generated JIT code.

**Comment by WilliamLCobb** [comment]
It works!!! I've been working on JIT all day and finally got it working on 32 bit devices. The bad news is the way I fixed it was to run it very slowly. For some reason slowing it down around 100 instructions per second fixed the random register clears. This is way too slow to play any games but it's huge progress. If I fix whatever is messing with the registers when the program is not slowed down, JIT will work 100%. It will even work on non-jailbroken iPhones

**Comment by WilliamLCobb** [comment]
Just got 60 fps on my iPad mini on a simple homebrew game (up from 16 fps). There's still problems but I'm going to talk to my professor tomorrow and hopefully he can help me sort it out.

**Comment by WilliamLCobb** [comment]
JIT is finished for 32bit devices. I hope it works on non-jailbroken phones. I'm going to work on getting it for 64bit arm devices next

**Comment by pmp174** [comment]
Yay !!!

**Comment by Anuskuss** [comment]
Wow, this is awesome. Your work here is incredible, and if I read that right you're saying that it might work even for non-jailbreakon phones? Simply amazing, keep it up!

**Comment by WilliamLCobb** [comment]
I hope so I really need someone to test it for me. If anyone has a 32bit phone (before the iPhone 6) and is able to side load the app and run it, please let me know

**Comment by Pepperoony** [comment]
It'd have to be a iPhone 5 or below, the iPhone 5s is 64 bit


--


Luke Rooney/ Director of Productions
luke.rooney@servitehs.org(mailto:luke_rooney@icloud.com)/(562)-360-4258(tel:(562)-360-4258)


Trinity Corporation
ServiteHS.org
1952 W La Palma Ave, Anaheim(x-apple-data-detectors://5/2)

**Comment by Anuskuss** [comment]
@WilliamLCobb My brother has a stock iPhone 5C. If he let me use it, I'll help you.

**Comment by Anuskuss** [comment]
I have an old iPad mini. Can I use it to help you?

**Comment by Anuskuss** [comment]
I have an old stock iPad mini. Can I use it to help you?

**Comment by WilliamLCobb** [comment]
Yea! Try building it from XCode and set the emulator engine to JIT in the settings. It it runs that means JIT will work on non-jailbroken devices

**Comment by Anuskuss** [comment]
#### Report
**Version:** iNDS 1.5.1 (*Debug*)
**Origin:** Git
**Device:** iPad mini - iOS 9.2.1 (jailed)
**Settings:** Default
**Tested Games:**
- Speed Racer - The Videogame [E]
- Pokemon - Black Version [E]
- Golden Sun - Dark Dawn [U]

All games are working in ~~Interpretor~~ *Interpreter* mode, but crashing with *JIT Recompiler (Beta)*.

#### Running with Xcode
2016-04-28 22:27:52.521 iNDS[24241:3587555] Loading ROM /var/mobile/Containers/Data/Application/FF0BB3C4-7BDC-4F0D-A1EC-6D276ED639A6/Documents/2317 - Speed Racer - The Videogame (E)(XenoPhobia).nds
2 cores detected
Init NDSSlot 2: MPCF Flash Card Device
MMU init
Mic Init
Microphone successfully inited.
DeSmuME 0.9.10 debug ARM-JIT NOSSE
SoftRast Initialized with cores=2
Init sound core
Already decrypted.

ROM game code: YYRP
ROM crc: E231B00B
ROM serial: NTR-YYRP-EUR
ROM internal name: SPEED RACER

Backup size: 3 Kbit
Loading cheatfile at path: /var/mobile/Containers/Data/Application/FF0BB3C4-7BDC-4F0D-A1EC-6D276ED639A6/Documents/Battery/2317 - Speed Racer - The Videogame (E)(XenoPhobia).dct
DEBUG_reset: 003DFC00
Using CFlash directory: 
Trying with 1 sectors/cluster:
FAT12: #clu=73266, fatlen=215, maxclu=4080, limit=4080
FAT12: too much clusters
FAT16: #clu=73124, fatlen=286, maxclu=65520, limit=65520
FAT16: too much clusters
FAT16: would be misdetected as FAT12
FAT32: #clu=72562, fatlen=567, maxclu=72576, limit=268435440
Using sector 6 as backup boot sector (0 = none)
Mic_Reset
Backup size: 3 Kbit
armcpu_setjitmode : Arm LJit
PageSize : 4096
CodeBuffer : start=0x40000000, size1=4194304, size2=268435456
Reserving CPU
sizeof(armcpu_t) = 256
sizeof(Decoded) = 52
sizeof(JitBlock) = 56
Loading /var/mobile/Containers/Data/Application/FF0BB3C4-7BDC-4F0D-A1EC-6D276ED639A6/Documents/2317 - Speed Racer - The Videogame (E)(XenoPhobia).nds was successful
SoftRast Initialized with cores=2
2016-04-28 22:27:59.281 iNDS[24241:3587555] Initiaizing GL
2016-04-28 22:28:00.163 iNDS[24241:3587905] Communications error: <OS_xpc_error: <error: 0x37007654> { count = 1, contents =
	"XPCErrorDescription" => <string: 0x37007854> { length = 22, contents = "Connection interrupted" }
}>
2016-04-28 22:28:00.254 iNDS[24241:3587555] Is multi: 0
2016-04-28 22:28:00.796 iNDS[24241:3587836] Starting controller discovery...

*Game doesn't crash now, but hangs on a black screen with* **? FPS**. *Also there's a blue* **(lldb)** *waiting for my input (in the Xcode output).*

#### ArmLJit.cpp
TEMPLATE static u32 cpuExecuteLJIT()
{
...
return opfun(); **Thread 9: EXC_BAD_ACCESS (code=2, address=0x40000000)**
}

**Comment by Anuskuss** [comment]
#### Report
**Version:** iNDS 1.5.1 (*Debug*)
**Origin:** Git
**Device:** iPad mini - iOS 9.2.1 (jailed)
**Settings:** Default
**Tested Games:**
- Speed Racer - The Videogame [E]
- Pokemon - Black Version [E]
- Golden Sun - Dark Dawn [U]

All games are working in ~~Interpretor~~ *Interpreter* mode, but crashing with *JIT Recompiler (Beta)*.

#### Running with Xcode
2016-04-28 22:27:52.521 iNDS[24241:3587555] Loading ROM /var/mobile/Containers/Data/Application/FF0BB3C4-7BDC-4F0D-A1EC-6D276ED639A6/Documents/2317 - Speed Racer - The Videogame (E)(XenoPhobia).nds
2 cores detected
Init NDSSlot 2: MPCF Flash Card Device
MMU init
Mic Init
Microphone successfully inited.
DeSmuME 0.9.10 debug ARM-JIT NOSSE
SoftRast Initialized with cores=2
Init sound core
Already decrypted.

ROM game code: YYRP
ROM crc: E231B00B
ROM serial: NTR-YYRP-EUR
ROM internal name: SPEED RACER

Backup size: 3 Kbit
Loading cheatfile at path: /var/mobile/Containers/Data/Application/FF0BB3C4-7BDC-4F0D-A1EC-6D276ED639A6/Documents/Battery/2317 - Speed Racer - The Videogame (E)(XenoPhobia).dct
DEBUG_reset: 003DFC00
Using CFlash directory: 
Trying with 1 sectors/cluster:
FAT12: #clu=73266, fatlen=215, maxclu=4080, limit=4080
FAT12: too much clusters
FAT16: #clu=73124, fatlen=286, maxclu=65520, limit=65520
FAT16: too much clusters
FAT16: would be misdetected as FAT12
FAT32: #clu=72562, fatlen=567, maxclu=72576, limit=268435440
Using sector 6 as backup boot sector (0 = none)
Mic_Reset
Backup size: 3 Kbit
armcpu_setjitmode : Arm LJit
PageSize : 4096
CodeBuffer : start=0x40000000, size1=4194304, size2=268435456
Reserving CPU
sizeof(armcpu_t) = 256
sizeof(Decoded) = 52
sizeof(JitBlock) = 56
Loading /var/mobile/Containers/Data/Application/FF0BB3C4-7BDC-4F0D-A1EC-6D276ED639A6/Documents/2317 - Speed Racer - The Videogame (E)(XenoPhobia).nds was successful
SoftRast Initialized with cores=2
2016-04-28 22:27:59.281 iNDS[24241:3587555] Initiaizing GL
2016-04-28 22:28:00.163 iNDS[24241:3587905] Communications error: <OS_xpc_error: <error: 0x37007654> { count = 1, contents =
	"XPCErrorDescription" => <string: 0x37007854> { length = 22, contents = "Connection interrupted" }
}>
2016-04-28 22:28:00.254 iNDS[24241:3587555] Is multi: 0
2016-04-28 22:28:00.796 iNDS[24241:3587836] Starting controller discovery...

*Game doesn't crash now, but hangs on a black screen with* **? FPS**. *Also there's a blue* **(lldb)** *waiting for my input (in the Xcode output).*

##### ArmLJit.cpp
TEMPLATE static u32 cpuExecuteLJIT()
{
...
return opfun(); **Thread 9: EXC_BAD_ACCESS (code=2, address=0x40000000)**
}

**Comment by DarknesGaming** [comment]
I really wish I had your skills @WilliamLCobb

**Comment by WilliamLCobb** [comment]
Thanks for posting that. Sadly it looks like KPP in iOS 9 is preventing JIT and it's unlikely we'll find a way around it. JIT will work on jailbroken 32bit iOS 9 devices and all 32bit iOS 8 devices but thats it. Maybe someone or I will find a way around this but for now JIT is very limited

**Comment by Razzile** [comment]
@WilliamLCobb unsure if it will work on iOS 9, but on OS X 10.11 it's possible to create executable pages ad-hoc by calling mmap with the flags MAP_ANON, MAP_PRIVATE and MAP_JIT, could be used to implement JIT on 64 bit devices? On OS X this doesn't require any entitlements either but I have a feeling it will need some sort of entitlement on iOS (if it even works). This method also seems to have far less restrictions than the usual `vm_allocate` method.

if this is what the JIT core already does then just ignore my message haha, I haven't yet looked at the source.

**Comment by WilliamLCobb** [comment]
It's not I'll definitely give that a shot in a minute and I'll let you know if it worked

**Comment by WilliamLCobb** [comment]
Wow! Thats jailed ios9?

**Comment by Razzile** [comment]
jailbroken, but if it works on stock OS X i guess it should work on jailed iOS 9

**Comment by WilliamLCobb** [comment]
You did it! JIT works on iOS 9

**Comment by WilliamLCobb** [comment]
Well it just crashed but I'm hopeful this will help me get it to work.

**Comment by Razzile** [comment]
@WilliamLCobb glad to hear it, but I have a feeling that this is a shortcoming on apple's part for not checking signing on the pages rather than a planned feature :disappointed:

**Comment by Anuskuss** [comment]
**Awesome!** @WilliamLCobb and @Razzile, do you mind sharing your experience with the PPSSPP team ([#8122](https://github.com/hrydgard/ppsspp/issues/8122))?

Plus now that you're working on the amd64 JIT, I have an jailbroken iPhone 6 8.4 (pre-KPP) if you need my help.

**Comment by WilliamLCobb** [comment]
I was wrong when I said it ran on jailed devices. I've been testing some more and I can't get it to work again. It might have defaulted to the interpreter without me knowing but I didn't save the output to check. 

I'm going to try to look for an exploit to break outside of the sandbox in order to protect memory pages for execution. I don't have a ton of experience in security but I have enough to get started. Anyway the speedup from JIT is kind of disappointing. I haven't optimized anything yet but it's only like 10% on my iPad mini and I thought it would be much higher. 

I'll definitely work on it this summer but until we find a way to get JIT running on jailed devices, working on the JIT implementation is pretty pointless.

**Comment by Razzile** [comment]
@WilliamLCobb I am now unjailbroken (funny what a day can do). I'll investigate JIT on jailed devices too and post any results

**Comment by Razzile** [comment]
my initial test i posted yesterday throws Invalid argument on jailed iOS 9.3.1 :/ still looking into it

**Comment by WilliamLCobb** [comment]
Woah really? I was getting a bad access with an error code of 2 meaning the privileges were off. What device is this on?

**Comment by WilliamLCobb** [comment]
I'm about to update my iPad and I'll check it out

**Comment by Razzile** [comment]
iphone 6+ 9.3.1

**Comment by Razzile** [comment]
I did some more testing to make sure it wasn't a fluke using shellcode that writes to stdout and it still worked.

![img](http://i.imgur.com/9p9PHg1.png)

**Comment by WilliamLCobb** [comment]
Yep it works! The key was removing write protection from the memory before executing. JIT is successfully running on my jailed iPad mini on iOS 9.3.1! I'll put out an update and I'll start working on arm64 support

**Comment by WilliamLCobb** [comment]
So JIT now works but there's a new problem. Since I have to protect the page for executing before JIT and protect it again for writing after every block, it's extremely slow (3fps slow). 

I'll see if I can figure out a way to only protect pages when necessary to speed it up

**Comment by WilliamLCobb** [comment]
So I fixed that but unfortunately there is another problem. JIT only works if you run iNDS from Xcode. If I try to launch it normally, it gives me a code signing error and crashes when I use mem_protect

**Comment by Razzile** [comment]
ah damn, thought that was too good to be true. It may still work on jailbroken devices I guess.

**Comment by DarknesGaming** [comment]
iNDS crashes a lot more often than normal. I figured out how to change the build configuration but I will do that when I get home.

**Comment by OkazakiTheOtaku** [comment]
I'm about to test JIT on my i6s+ on 9.3. Will report back.

**Comment by WilliamLCobb** [comment]
I've been looking a ton into the iOS signing mechanisms and it looks like it's going to be near impossible to implement for jailed devices. iOS checks every single page before loading it back into memory on a fault. There used to be a bug in earlier iOS version that let you protect pages once per process but it looks like it's been patched

**Comment by OkazakiTheOtaku** [comment]
Welp I can't even enable it, the app won't let me
Lol

**Comment by WilliamLCobb** [comment]
lol

change `EMU_setCPUMode((int)[[NSUserDefaults standardUserDefaults] integerForKey:@"cpuMode"] + 1);`
to
`EMU_setCPUMode(2);`

**Comment by OkazakiTheOtaku** [comment]
Thanks, will report back

**Comment by OkazakiTheOtaku** [comment]
Didn't work, still saying that "JIT is not yet available for your device."

**Comment by OkazakiTheOtaku** [comment]
I realize now that it was just supposed to run JIT all the time even when its not 'enabled'.

It locks up and outputs this
http://imgur.com/gdjckKS

**Comment by Razzile** [comment]
>I've been looking a ton into the iOS signing mechanisms and it looks like it's going to be near impossible to implement for jailed devices.

the MAP_JIT method is *supposed* to be Apple's go-to method when a developer needs to run dynamic code. the Safari/JavascriptCore uses it to compile javascript on the fly. safari used to need a `dynamic-codesigning` entitlement to use it though, but I haven't seen this entitlement since iOS 4 or 5.

Maybe apple just decided with iOS 9 to cut it out.

If it's any conciliation there's [this project](https://github.com/kpwn/921csbypass) which could allow JIT to work up to iOS 9.2.1. 

I'll be looking into getting it to work on iOS 9.3.1 but I have exams starting this week so I doubt i'll get anywhere

P.S @OkazakiTheOtaku I'm stealing your avatar 😁

**Comment by OkazakiTheOtaku** [comment]
Interesting.
Well I have a 5s and a 6s+ to test on if you need any testers. Currently they're both on 9.3 but I'll jailbreak at least one of them when it comes out.

@Razzile lol okay! :) always awesome to see a fellow Miku fan

**Comment by DarknesGaming** [comment]
Always waiting for iOS 9.3 jailbreak. People think that there won't be a 9.3 jailbreak but there will be a 9.3.2 jailbreak instead due to the bugs found in 9.3.

**Comment by OkazakiTheOtaku** [comment]
I don't know. I'm on 9.3.0 on my daily driver and 9.3.1 on my secondary driver (by accident, I'm stupid) and I don't plan on budging. I'm hoping it comes soon.

**Comment by DarknesGaming** [comment]
A jailbreak for iOS 9.3-9.3.2 should be released before June 13, which is when the WWDC event happens.

**Comment by DarknesGaming** [comment]
I just went through the entire iNDS source code again. This app is using GNU Lightning JIT 1.2 built in 2004. The latest one (2.1.0) was made last year, and I still have no clue how to build it. If I do figure it out, hopefully it will be faster than 1.2.
Also, which files use TinyXML? I wanted to upgrade it to TinyXML2, which uses less memory, is faster, and with far less memory allocations. You can look it up on Google.

**Comment by DarknesGaming** [comment]
Also, how do you build the GNU lightning Git? I can't figure it out, and I did look it up and got atuck on the first step:

**Comment by DarknesGaming** [comment]
I just figured out how to build GNU Lightning Git and succeeded on my first try! THIS IS AWESOME!
What I don't know is how @WilliamLCobb managed to build it to get those arm, sparc and i386 folders with more c++ files inside.

**Comment by iSuns9** [comment]
This is probably a bad idea but would it be possible to use an interpreter for the JIT generated code and still be faster than the usual interpreter? This should work even on jailed 64bit devices. But anyway why can't JIT be activated on jailbroken 32bit devices? This should be easy, shouldn't it? Because who needs an emulator that can only be bootstrapped by XCode?

**Comment by Razzile** [comment]
JIT does work on 32 devices afaik, it's 64 bit devices that it doesn't work on

**Comment by DarknesGaming** [comment]
We might have to build a 64-bit JIT for this, but I'm not sure if it will be Lightning GIT or ARM JIT. Only time will tell.

**Comment by Razzile** [comment]
32 bit JIT should run fine on 64 bit devices. The problem is that in iOS 9 on 64 bit devices apple have many different security checks (such as KPP) that prevent JIT code being executed.

**Comment by DarknesGaming** [comment]
We should probably build 64 bit JIT then. Hopefully, it works because that will be great for iNDS 2.0.0.
Also, I finally figured out how to compile Lightning JIT, which is very good.

**Comment by WilliamLCobb** [comment]
I'm going to try and clear everything up because there are a lot of questions.

## GNU Lightning:
--------------------------------------------------------------------------------
iNDS uses GNU Lightning 1.2 to produce JIT code. GNU Lightning 1.2 was written before the arm64 instruction set so it does not produce the right instructions for arm64. It works perfectly on 32 bit devices. 
GNU Lightning 2.1 supports arm64 but using it is not as easy as just compiling it. The syntax is different and much of the JIT code will need to be rewritten in order to get it to work.

## Code Signing:
--------------------------------------------------------------------------------
Even though JIT now works on 32-bit devices, there are still hurdles for jailed devices. Allowing apps to generate and run their own code is a serious security vulnerability and Apple goes to great lengths to prevent JIT. Before any JIT code is executed, iNDS has to edit virtual memory pages in order to receive permission to jump to the JIT code stored there. Virtual pages are allocated and given to iNDS by the kernel with read and write privileges. iNDS has to add execution privileges before and JIT code stored in those pages can run. The problem is the iOS kernel checks every single page loaded into memory to make sure it has the right privileges. When we load the first JIT block, the kernel sees iNDS has changed the privileges of the page and crashes, or it removes the privileges and crashes when the JIT runs. On jailbroken devices, this isn't a problem because the kernel has been edited to remove these checks. However, iOS 9 introduced KPP (Kernel Patch Protection), which means it's nearly impossible for anything to edit (Patch) any part of the kernel. 

One way around this is getting the JIT entitlement for iNDS. Mobile safari is the only app on iOS that has this entitlement. It uses JIT to compile javascript in order to run much faster. There have been hacks in the past to get the JIT entitlement on an app or to trick the kernel into giving executable pages. This is what [cs_bypass](https://github.com/kpwn/921csbypass) does. I haven't had time to test that out, but theoretically it could give iNDS executable pages to use and fix all of the above problems.

## Value of JIT:
--------------------------------------------------------------------------------
One important thing that needs to be discussed is the actual value of JIT. Only ~34% of the actual execution time of iNDS is spent emulating the CPU. Thats a big chunk, but certainly not very close to 100%. The other 66% is spent emulating the GPU. JIT only speeds up the CPU emulation and the JIT implementation we have now barely provides a speedup in the overall execution time. Even if it was twice as fast as the interpreter, iNDS would only be a 17% faster. So if you're running at 30 fps, turning JIT on will only bring you to 35 fps.

If there's anything that needs focus, it's the GPU. The 66% percent of time emulating the GPU could be significantly reduced. Running it on an actual GPU could easily double or triple the framerate, turning that 30 fps in to 60 or 90. It would also allow for much better graphic performance and higher resolution characters (sprites would still be low quality).

**Comment by DarknesGaming** [comment]
I'm back with good research.
iOS 8 can run on devices with CPUs as early as the Apple A5, which runs on ARM architecture. Why iNDS has the old i386 JIT code, I have no idea.
The Apple A7 (iPhone 5S) is on a 64-bit architecture, known as AArch64 (ARMv8). We can build those and port for Lightning JIT, if possible.

**Comment by DarknesGaming** [comment]
I just read the very long comment that @WilliamLCobb created. I somehow did not notice it on my phone, but I did on my actual computer. We really should focus on making the GPU actually do something, and since all the iPhones use PowerVR (4S and up), that makes the whole problem much easier to solve.

**Comment by DarknesGaming** [comment]
Also, try building GNU Lightning GIT with the host, build and target set to arm-apple-darwin like this:
./configure --host=arm-apple-darwin --build=arm-apple-darwin --target=arm-apple-darwin
Both Mac OS X and iOS run on the Darwin kernel, so this command I just typed up should compile the JIT code to work exclusively with the Apple Darwin kernel. Hopefully, the KPP will allow iNDS to work using such a code.
Speaking of JIT, how do you even port this? I got to that part of the manual, but I don't have the lightning.m4 and configure.in files needed for this.
@WilliamLCobb you probably managed to do this. Can you teach me on how to do this?

**Comment by DarknesGaming** [comment]
I am really hoping @WilliamLCobb notices my pull requests. I am trying to be helpful.

**Comment by iSuns9** [comment]
@WilliamLCobb I still haven't downloaded and built the iNDS source. Without JIT I get 34-35 fps on my iPhone 5C in Pokemon Diamond. So does your long comment mean I wont even get 50fps with JIT?

**Comment by DarknesGaming** [comment]
Let me answer that for him:
OBVIOUSLY. THE 5C IS JUST AS FAST AS THE NORMAL 5.

**Comment by WilliamLCobb** [comment]
@iSuns9 was asking about how much JIT would speed up emulation not about how fast it was compared to the 5. Currently I don't think you would get 50 fps. JIT will definitely increase fps but I don't think It will make a huge difference until it's optimized for the iPhone. I do think there's hope through other ways that the 5c could eventually run at 50 or 60 fps

**Comment by DarknesGaming** [comment]
I did research on the iPhone 5C and the iPhone 5. Both of them use the same CPU, same GPU, and same memory type and speed. Other than the colors, there really isn't a difference in terms of performance.
There is a way that we can optimize JIT. Still working on configuring and porting though.

**Comment by DarknesGaming** [comment]
I'm still working on iNDS 2.0.0, and every bug that was fixed in the master branch after May 3 still isn't fixed in the 2.0 branch. I made a pull request about that.
So far, I am currently working on upgrading TinyXML, 7-zip, minizip, and a bunch of other stuff. I really need help.

**Comment by DarknesGaming** [comment]
I'm going to test the ArmThreadedInterpreter, ArmCJit and ArmLJit on DeSmuME 0.9.11. How do people even manage to build those?

**Comment by DarknesGaming** [comment]
What is void*4?

**Comment by Razzile** [comment]
@DarknesGaming what context? a check if sizeof(void*) == 4 is to determine whether or not the process is running in 32 bit or 64 bit because sizeof(void*) in 64 bit environments = 8

**Comment by DarknesGaming** [comment]
Thank you for the valuable information @Razzle. That explains why JIT still isn't available for 64-bit devices, and it shows how much of a noob I am with C++ and Objective-C code.

**Comment by DarknesGaming** [comment]
I just looked into the GNU Lightning 1.2 source code, and compared it to the new 2.1 code. That explains how @WilliamLCobb managed to port GNU Lightning in the first place, but doesn't explain how the ArmLJit.cpp file was created.

**Comment by Anuskuss** [comment]
Hey @WilliamLCobb, I know we need better Renderer but have you looked at the (newly implemented) **IR Interpreter** (Intermediate Representation) from the PPSSPP team [#8725](https://github.com/hrydgard/ppsspp/pull/8725)?

> **unknownbrackets**
We have recently implemented (mostly **@​hrydgard**) a new, experimental interpreter - it still doesn't use jit, but it pre-processes the MIPS code to generate something that is faster to interpret. In other words, it does the 25% of what jit does, the part that is allowed on Apple devices.

**Comment by Anuskuss** [comment]
Hey @WilliamLCobb, I know we need a better Renderer but have you looked at the (newly implemented) **IR Interpreter** (Intermediate Representation) from the PPSSPP team [[#8725](https://github.com/hrydgard/ppsspp/pull/8725)]?

> **unknownbrackets**
We have recently implemented (mostly **@​hrydgard**) a new, experimental interpreter - it still doesn't use jit, but it pre-processes the MIPS code to generate something that is faster to interpret. In other words, it does the 25% of what jit does, the part that is allowed on Apple devices.

**Comment by DarknesGaming** [comment]
@Anuskuss not to sound like a critic or anything, but do you think that would work with DeSmuME? DeSmuME already has a JIT compiler, but to enable that is by getting into the terminal of Apple devices to use --cpu-mode=1.

**Comment by Anuskuss** [comment]
@DarknesGaming We can't use the JIT compiler on iOS9 so this was a suggestion to look at an alternative.

**Comment by Razzile** [comment]
i've got JIT working on non jailbroken devices *again* woo.

thanks to qwertyoruiop for telling me how. 

basically you do the same as what i was doing above (map a page as RW, copy JIT code to it, map it as RX) but before then, you need to call `ptrace(0,0,0,0);` this disables codesigning on pages in the process including the newly created page, allowing it to run.

![image](https://cloud.githubusercontent.com/assets/5256381/15419890/36052b82-1e61-11e6-8bc6-ca2fcbbae02e.png)
The above will output hello, world to stdout. works when launched from xcode or springboard.
note the 1st line `syscall(26, 0, 0, 0, 0);` is just the same as calling `ptrace(0, 0, 0, 0);`

One other thing to note is that qwertyoruiop mentioned that this wouldn't work if the binary was signed for distribution. It would only work when signed with a developer certificate. I'm not sure how iNDS is distributed for jailed devices and I'm no expert with certs because I don't have an enterprise cert to check anything so you're on your own for that part. 

Absolute worst case scenario just tell people to build it themselves if they want JIT.

**Comment by Fadi5555** [comment]
@Razzile wow this is a great news indeed.
is that this method will include other projects like ppsspp and retroarch?
Also is this method will work on both jailed and jailbroken devices?
Thanks a lot.

**Comment by DuIslingr** [comment]
Based on what he wrote. It should work fine on jailbroken devices. Jailed however requires it to be signed with a dev certificate. And I believe that the other projects mentioned above can use this as well.

**Comment by WilliamLCobb** [comment]
Wow that's awesome! I'll add that in a second and see how it goes. 

The last thing we need is information on upgrading GNU lightning to 2.1. The syntax isn't that different from 1.2 but I've found no documentation for 2.x other than a few examples. No change log in the source or anything for upgrading from 1.2.

**Comment by WilliamLCobb** [comment]
The other alternative if to keep using 1.2 and add a custom arm64 support. I'd be surprised if someone hasn't already done this but I haven't been able to find it. 

Also I know 64bit processors are able to execute 32bit instructions but I don't know if it's possible to switch mid execution on iOS. If we could enable this somehow, very little work would be needed for JIT on all devices.

**Comment by DarknesGaming** [comment]
I've been trying to get GNU Lightning 2.1 to function, but all the guides and manuals haven't been updated in over 10 years.
I'll just have to come up with my own way to do this.
@Razzile that picture you uploaded. Which lines contain that code, and on which file?
I need to start reading those Apple coding guides I downloaded from iBooks.

**Comment by Pepperoony** [comment]
Just so I can it out quick before I forget (this probably isn't the place to put this) but the 1.5.3 update started pumping any sound from the volume bumper from the top of my iPhone instead from the speakers. I don't have volume bumpers turned on or microphone and I have no idea how to fix it!


--


Luke Rooney/ Director of Productions
luke.rooney@servitehs.org(mailto:luke_rooney@icloud.com)/(562)-360-4258(tel:(562)-360-4258)


Trinity Corporation
ServiteHS.org
1952 W La Palma Ave, Anaheim(x-apple-data-detectors://5/2)

**Comment by Razzile** [comment]
All the code is in the screenshot.

#include <sys/mman.h> for mmap(). shell is the byte code representation of syscall(SYS_write, stdout, arg1, arg2) [or equivalent] to write to stdout

**Comment by WilliamLCobb** [comment]
It works! I wish we could get the JIT flag somehow because the memory has to be protected twice every time a block is compiled (Only JIT marked pages can be write and exec).
 
One weird thing is after adding the syscall my iPad is completely unusable until I restart it after I close iNDS. Buttons stop working and it stops responding to events. If I tap a button or icon the animation happens but nothing else. It also refuses to unlock after I lock it telling me my passcode is incorrect.

Very odd, I'll upload the change and see if anyone else has similar problems

**Comment by Razzile** [comment]
@WilliamLCobb the JIT flag requires the entitlement `dynamic-codesigning`. if you can somehow sign your app with that entitlement (I tried on my free dev account and xcode just rejected it) you can use MAP_JIT. obviously on jailbroken devices you can just pseudosign with ldid and a plist containing something like this:

```plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>dynamic-codesigning</key>
        <true/>
        <key>get-task-allow</key>
        <true/>
    </dict>
</plist>
```
(not sure if get-task-allow is needed but included for good measure)

**Comment by Razzile** [comment]
also @WilliamLCobb that device freezing happened to me too but only once out of the 5 or so times i was testing JIT. it's very odd indeed. you could try calling ptrace(-1, 0, 0, 0) which is an invalid flag, hopefully ptrace would just error but the system would still disable codesigning first

**Comment by DarknesGaming** [comment]
I would try this, but after @WilliamLCobb's device going nuts I don't think that would be a good idea. I'm assuming his iPad is 32-bit because it's still not available for 64-bit devices yet.
Also, if you have a 9.7" iPad Pro, don't update to iOS 9.3.2. It bricks your device and iTunes can't fix it, and creates error 56.

**Comment by DuIslingr** [comment]
@DarknesGaming Lol you dont have to worry about the 9.3.2 update. Apple pulled it. Second time they have pulled an update and that last one was recent too. Regarding your comment about it not being on 64 bit device, I don't know what you mean since @Razzile  has mentioned above that hes been doing his testing on an iPhone 6+ 9.3.1 

Summary of JIT status on iOS 9
Jailbroken 32 bit devices have JIT functioning(due to the jailbreak on those devices disabling Rootless(KPP). However they didn't do this on ARM64 for reasons I can't remember. Now based on what I have read regarding this issue, if they get JIT to function 100% on jailed iOS 9 devices then this should also solve jailbroken ARM64 devices not being able to use JIT. 

So as of right now thanks to Razzile  we are now a step closer to getting it working. Just need to work out bugs with it.

**Comment by DarknesGaming** [comment]
We could add an arm64 backend to GNU Lightning, but that means using the 2.0-2.1 source code and editing the ArmLJit.cpp file. Plus, the GNU Lightning source code refers "arm64" as "aarch64" instead.

**Comment by DarknesGaming** [comment]
We could add an arm64 backend to GNU Lightning, but that means using the 2.0-2.1 source code and editing the ArmLJit.cpp file. Plus, the GNU Lightning source code refers "arm64" as "aarch64" instead.

**Comment by AMorgaut** [comment]
For your information Apple recently announced changes in the way JSCore & its JIT are managed in iOS 10

> “Hardened WebKit JIT Mapping” protects the Safari Browser on iOS. Most other devices, including Amazon Kindle, compile JavaScript code in beforehand. But iOS assures quick JavaScript execution by compiling JavaScript just before execution, (JIT). But there is a downside to JIT implementation- it demands far much less strict code-signing, thereby making the system vulnerable.
> Apple addressed the issue in the latest security measures on iOS. **Apple patched the JIT security concerns by dedicating specific areas of application memory to particular tasks**. Unlike before, JavaScript is now not allowed to run code in raw data storage in iOS 9 and previous versions; raw data storage is highly vulnerable to attacks. **On the iOS 10, the compiled JavaScript code runs on execution-particular memory**. Once the code is running on this memory, it cannot be changed, thus preventing on-the-fly code alterations attacks.

http://www.opptrends.com/2016/08/apple-security-measures-on-ios-detailed-at-the-black-hat-conference/

**Comment by Razzile** [comment]
That doesn't look like good news 😔 fortunately the JIT method mentioned above isn't used by safari but by debugserver instead. Hopefully it remains unchanged

**Comment by druidwarrior** [comment]
Talk to the devs of ppsspp I am currently using iOS 9.3.3 and they got hit working on ppsspp.

**Comment by druidwarrior** [comment]
iPhone 6 Plus by the way

**Comment by Razzile** [comment]
I'm not sure the current state of JIT. I've been away from iNDS for a while, but I have been messing around with mmap again on iOS 9.3.3.

if your app has the `dynamic-codesigning` entitlement, you can use the MAP_JIT flag in mmap to have memory as truly RWX (i think i mentioned this before actually).

not sure if this is useful for iNDS or if the current system (interpreted or anything else) is okay as it is

**Comment by Razzile** [comment]
here I am, bringing up dead issues again hehe

I noticed in coolstar's latest iOS 11 jailbreak code [this commit](https://github.com/coolstar/async_awake-fun/commit/61eaed15f9f27f36d07437a441576f306ad1de68) seems to be make code run like it's being debugged, I talked about loong ago. If that's the case then I see no reason why JIT wouldn't work with this jailbreak. I haven't tested yet though

**Comment by kprinssu** [comment]
iNDS displays a black screen with the controller overlay and any subsequent launches results in the same black screen and the app is unusable. My fix is to do a hard reset (restart my iPhone). My guess is that iNDS has 32bit instructions that is not supported on iOS 11 (please do correct me).
